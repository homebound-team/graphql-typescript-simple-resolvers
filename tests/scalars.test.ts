import { createTestSchema, runPlugin } from "./utils/test-helpers";

describe("Scalar Types", () => {
  const schemaWithScalars = `
    scalar Date
    scalar DateTime
    scalar CustomId

    type User {
      id: CustomId!
      name: String!
      birthday: Date
      lastLogin: DateTime
    }

    type Query {
      user(id: CustomId!): User
      usersBornAfter(date: Date!): [User!]!
      usersLoggedInAfter(datetime: DateTime!): [User!]!
    }

    input UserInput {
      name: String!
      birthday: Date
      customId: CustomId
    }

    type Mutation {
      createUser(input: UserInput!): User!
    }
  `;

  it("generates scalar types in resolver interfaces", async () => {
    const schema = createTestSchema(schemaWithScalars);
    const code = await runPlugin(schema, {
      scalars: {
        Date: "Date",
        DateTime: "Date",
        CustomId: "string",
      },
      mappers: {},
      enumValues: {},
    });

    // Should include scalar types in resolver interface
    expect(code).toContain(`export interface Resolvers {
  Query: QueryResolvers;
  Mutation: MutationResolvers;
  User?: UserResolvers;
  Date: GraphQLScalarType;
  DateTime: GraphQLScalarType;
  CustomId: GraphQLScalarType;
}`);
  });

  it("handles scalars in field types", async () => {
    const schema = createTestSchema(schemaWithScalars);
    const code = await runPlugin(schema, {
      scalars: {
        Date: "Date",
        DateTime: "Date",
        CustomId: "string",
      },
      mappers: {},
      enumValues: {},
    });

    // User interface should use scalar mappings
    expect(code).toContain(`export interface User {
  id: string;
  name: string;
  birthday: Date | null | undefined;
  lastLogin: Date | null | undefined;
}`);
  });

  it("handles scalars in query arguments", async () => {
    const schema = createTestSchema(schemaWithScalars);
    const code = await runPlugin(schema, {
      scalars: {
        Date: "Date",
        DateTime: "Date",
        CustomId: "string",
      },
      mappers: {},
      enumValues: {},
    });

    // Query args should use scalar types
    expect(code).toContain(`export interface QueryUserArgs {
  id: string;
}`);
    expect(code).toContain(`export interface QueryUsersBornAfterArgs {
  date: Date;
}`);
    expect(code).toContain(`export interface QueryUsersLoggedInAfterArgs {
  datetime: Date;
}`);
  });

  it("handles scalars in input types", async () => {
    const schema = createTestSchema(schemaWithScalars);
    const code = await runPlugin(schema, {
      scalars: {
        Date: "Date",
        DateTime: "Date",
        CustomId: "string",
      },
      mappers: {},
      enumValues: {},
    });

    // Input types should use scalar mappings
    expect(code).toContain(`export interface UserInput {
  name: string;
  birthday?: Date | null | undefined;
  customId?: string | null | undefined;
}`);
  });

  it("defaults to 'any' for unmapped scalars", async () => {
    const schema = createTestSchema(schemaWithScalars);
    const code = await runPlugin(schema, {
      scalars: {
        // Only map some scalars, leave others unmapped
        Date: "Date",
      },
      mappers: {},
      enumValues: {},
    });

    expect(code).toContain(`export interface User {
  id: CustomId;
  name: string;
  birthday: Date | null | undefined;
  lastLogin: DateTime | null | undefined;
}`);
  });

  it("snapshots scalar type generation", async () => {
    const schema = createTestSchema(schemaWithScalars);
    const code = await runPlugin(schema, {
      scalars: {
        Date: "Date",
        DateTime: "Date",
        CustomId: "string",
      },
      mappers: {},
      enumValues: {},
    });
    expect(code).toMatchSnapshot();
  });
});
