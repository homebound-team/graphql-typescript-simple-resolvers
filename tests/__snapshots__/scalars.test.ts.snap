// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Scalar Types snapshots scalar type generation 1`] = `
"import { GraphQLResolveInfo, GraphQLScalarType } from "graphql";
import { Context } from "./context";

export interface Resolvers {
  Query: QueryResolvers;
  Mutation: MutationResolvers;
  User?: UserResolvers;
  Date: GraphQLScalarType;
  DateTime: GraphQLScalarType;
  CustomId: GraphQLScalarType;
}

export type UnionResolvers = {};

export interface QueryResolvers {
  user: Resolver<{}, QueryUserArgs, User | null | undefined>;
  usersBornAfter: Resolver<{}, QueryUsersBornAfterArgs, readonly User[]>;
  usersLoggedInAfter: Resolver<{}, QueryUsersLoggedInAfterArgs, readonly User[]>;
}

export interface MutationResolvers {
  createUser: Resolver<{}, MutationCreateUserArgs, User>;
}

export interface UserResolvers {
  id: Resolver<User, {}, string>;
  name: Resolver<User, {}, string>;
  birthday: Resolver<User, {}, Date | null | undefined>;
  lastLogin: Resolver<User, {}, Date | null | undefined>;
}

type MaybePromise<T> = T | Promise<T>;
export type Resolver<R, A, T> = (root: R, args: A, ctx: Context, info: GraphQLResolveInfo) => MaybePromise<T>;

export type SubscriptionResolverFilter<R, A, T> = (
  root: R | undefined,
  args: A,
  ctx: Context,
  info: GraphQLResolveInfo,
) => boolean | Promise<boolean>;
export type SubscriptionResolver<R, A, T> = {
  subscribe: (root: R | undefined, args: A, ctx: Context, info: GraphQLResolveInfo) => AsyncIterator<T>;
};

export interface QueryUserArgs {
  id: string;
}
export interface QueryUsersBornAfterArgs {
  date: Date;
}
export interface QueryUsersLoggedInAfterArgs {
  datetime: Date;
}
export interface MutationCreateUserArgs {
  input: UserInput;
}
export interface User {
  id: string;
  name: string;
  birthday: Date | null | undefined;
  lastLogin: Date | null | undefined;
}

export interface UserInput {
  name: string;
  birthday?: Date | null | undefined;
  customId?: string | null | undefined;
}
"
`;
